Build and test on the glinet-b1300

rmelotte opened the pull request:
- build for the glinet-b1300
- test on the glinet-b1300
- collect logs and data and save them as artifacts
- other improvements to the CI

rmelotte committed:
tools: docker: openwrt: explicitly compile only prplMesh

The previous Dockerfile ran a plain 'make', relying on the
configuration to only compile prplMesh and its dependencies.

This is however not possible for all the targets. For the glinet-b1300
for example, the default is to build u-boot, which we cannot and don't
want to compile.

Replace the make invocation with to compile prplmesh explicitly.  Note
that we still want to fill the config file, and run "make defconfig",
because in some cases the prplMesh Makefile can depend on some config
options.

Also "cat" the configuration file before compiling, so that we see
what the result of "make defconfig" was.
This is useful to debug failing docker builds. For example, maybe we
specified a non-existant symbol, or the prpl feed had a breaking
change. We have a chance to spot that by looking at the config
file.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: build and test for the glinet-b1300

Add a new target device: glinet-b1300.

CI: Build the ipk, and test on the new target.

Also change the test script to properly stop prplMesh at the
end. There already is another script that kills prplMesh before
running the tests, but here we give it a chance to stop properly.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: openwrt: increase the status test timeout

When running the status test, we only wait for a specified time for
prplMesh to become operational. This time was set to 10 seconds, which
can be a little bit too small. As a consequence, the test would fail
from time to time, while re-running it would make it succeed.

Set the timeout to 30 seconds, to give prplMesh more time to restart.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

dob71 committed:
tools: docker: usability tweaks

- use sed instead of the '--exclude' option of git-describe, because
  it's currently not supported by git in some distros (Ubuntu 16.04)
- added command line option to specify docker build options. This can
  be used to pass the "--no-cache" option to docker for example (it
  might be necessary to force a rebuild since OpenWrt does not lock in
  versions of the feeds and each build therefore might be different
  even if openwrt commit SHA is still the same)
- move the check of TARGET_DEVICE after the loop, so it also detects a
  missing -d option
- fixed a few typos

Signed-off-by: Denis Bakin <denis.bakin@minim.co>
Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

dob71 committed:
tools: ci: add test run logs

To make it easier to debug failing tests, get the logs folder from the
target, compress it and save it as an artifact.

Also when restarting prplMesh, delete the old logs first.

Signed-off-by: Denis Bakin <denis.bakin@minim.co>

dob71 committed:
tools: ci: collect diagnostic data

Collecting diagnostic data is important to understand why a particular
test failed on some device.

Collect device information before running the status test:
- OS release date info and version
- uci configuration for prplmesh, network and wireless
- ifconfig output

Collect info after running the status test:
- output of bml_conn_map and bml_nw_map_query
- kernel messages

Also store all all the information in a tarball, and save it as an artifact.

Signed-off-by: Denis Bakin <denis.bakin@minim.co>

tomereli commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -81,8 +81,9 @@ RUN echo "src-git prpl ${PRPL_FEED}" >> feeds.conf.default \
     && make defconfig \
-    && make -j$(nproc) V=sc \
-    && make -j$(nproc) \
+    && cat .config \
```
Grammar mistake in commit message - 
>Also "cat" the configuration file before compiling, so that we now what the result of "make defconfig" was.

Should be 

> Also "cat" the configuration file before compiling, so that we see what the result of "make defconfig" was.

rmelotte replied:
Done.

tomereli commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -81,8 +81,9 @@ RUN echo "src-git prpl ${PRPL_FEED}" >> feeds.conf.default \
     && make defconfig \
-    && make -j$(nproc) V=sc \
-    && make -j$(nproc) \
+    && cat .config \
```
I'm not sure that dumping the entire .config (is it the whole .config or just the diff?) is the correct thing to do here, we better save it as an artifact like done for logs (I guess we don't dump the whole logs when things go well, do we?)

rmelotte replied:
Although it's being dumped when building the dockerfile, it won't be visible directly in the CI log, so I think it's still ok.
It will only be visible in the build.log file, which already contains a very verbose build output.

It's currently not possible to add the file as an artifact, because the CI does not run _inside_ the container (if we really want to have it as an artifact, we could still copy it out of the container).

tomereli commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -81,8 +81,9 @@ RUN echo "src-git prpl ${PRPL_FEED}" >> feeds.conf.default \
-    && make -j$(nproc) V=sc \
-    && make -j$(nproc) \
+    && cat .config \
+    && make package/feeds/prpl/prplmesh/compile -j$(nproc) V=sc \
```
What does `make V=sc` do? I think it should be documented in a comment

rmelotte replied:
`V` controls the verbosity, `sc` is the most verbose. We use that to know what's happening when building. 
For reference, it's documented here: https://openwrt.org/docs/techref/buildroot#warnings_errors_and_tracing
I will add a comment at the end.

tomereli commented on tools/docker/builder/openwrt/Dockerfile:
```diff
@@ -81,8 +81,9 @@ RUN echo "src-git prpl ${PRPL_FEED}" >> feeds.conf.default \
+    && make package/feeds/prpl/prplmesh/compile -j$(nproc) V=sc \
+    && make package/feeds/prpl/prplmesh/compile -j$(nproc) \
     && make package/feeds/prpl/prplmesh/clean
     # For some reason, make has to be run twice because
```
With this change, do we still need to run make twice? shouldn't we simply clean prplmesh then run make? why the need to compile it twice, then clean it, then run full make? I guess I'm missing something here.

rmelotte replied:
We previously had to run it twice because the build system would always try to rebuild the dependencies on the second run (but not on the third). I will check if it's still the case with this change.

Running make then clean is just a way to have the dependencies (but *not* prplmesh itself) already compiled inside the image. I don't know any way to build only the dependencies of a package in OpenWrt.

rmelotte replied:
Yes, running make twice is still needed.
We can probably avoid it by using CONFIG_AUTOREBUILD, but that can probably be done in another PR, because it needs testing.

tomereli commented on .gitlab-ci.yml:
```diff
@@ -72,8 +72,19 @@ build-for-turris-omnia:
   variables:
     TARGET_DEVICE: "turris-omnia"
 
+build-for-glinet-b1300:
```
I can't put the comment in the correct line but I think the previous commit should have added the .config to the artifact paths in this file.

rmelotte replied:
Is there a specific reason to keep it? It's already in the build.log file (see previous comment).

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -63,9 +63,15 @@ main() {
                     *)
                         err "Unknown target device: $TARGET_DEVICE"
-                        info "Currently supported targets are: turris-omnia"
+                        info "Currently supported targets are:"
```
It makes sense to put the supported targets in a string and iterate those here.

rmelotte replied:
Done.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -24,3 +24,5 @@ done
 exit 1
 EOF
 
+echo "Stopping prplMesh"
```
This is really not needed since you are doing `restart` in the test.
Also, please add the `-d` flag to the restart command - this will delete the logs from the previous run.

rmelotte replied:
We need to stop it at the end of the test to avoid the multiple target from interfering with one another.
Now this one is not strictly needed because we will kill all the instances before the tests starts, but it gives it a chance to stop more gracefully.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -14,7 +14,7 @@ TARGET="$1"
 ssh "$TARGET" <<"EOF"
 /opt/prplmesh/prplmesh_utils.sh restart
-TIMEOUT=10
+TIMEOUT=30
```
It makes more sense to add the TIMEOUT as a script argument IMO.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -57,6 +59,7 @@ main() {
     while true; do
         case "$1" in
             -h | --help)               usage; exit 0; shift ;;
+	    -b | --build-options)      BUILD_OPTIONS="$2"; shift ;;
```
Indentation looks wrong here

rmelotte replied:
Done. 

Also fixed: this commit missed passing BUILD_OPTIONS to build_image() (i.e. BUILD_OPTIONS was unused).

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -103,7 +107,7 @@ main() {
     export OPENWRT_VERSION
     export TARGET
-    PRPLMESH_VERSION="$(git describe --always --dirty --exclude '*')"
+    PRPLMESH_VERSION="$(git describe --always --dirty | sed -e 's/.*-g//')"
```
I would add a comment here saying what this sed command is doing, and change the commit message to mention that instead of exclude which is not supported in some git versions, we are using sed.

rmelotte replied:
Done.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -13,6 +13,7 @@ TARGET="$1"
 
 # the first start currently never succeeds, we need to restart it first.
 ssh "$TARGET" <<"EOF"
+rm -rf /tmp/beerocks/logs/*
```
Instead of `rm -rf` please use the `-d` option in the restart command:
```/opt/prplmesh/prplmesh_utils.sh restart -d```

rmelotte replied:
Done.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -11,7 +11,25 @@ fi
+echo
+echo "UCI Configuration (network and wireless):"
+uci export network
+uci export wireless
```
`uci export prplmesh` is missing

rmelotte replied:
Done.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -11,7 +11,25 @@ fi
+uci export wireless
+echo
+echo "ifconfig output:"
+ifconfig
```
I would also add wireless information with `iw`

rmelotte replied:
Done.

tomereli commented on tools/docker/tests/openwrt/test_status.sh:
```diff
@@ -24,15 +42,29 @@ for _ in $(seq 1 "$TIMEOUT") ; do
+echo "Output of bml_conn_map:"
+/opt/prplmesh/bin/beerocks_cli -c bml_conn_map 2>&1
+echo "Output of bml_nw_map_query:"
+/opt/prplmesh/bin/beerocks_cli -c bml_nw_map_query 2>&1
```
What is `bml_nw_map_query`?

rmelotte replied:
It's at least being used by the analyzer, and from what I recall its output can be different than `bml_conn_map`.
I don't know to what extent it's relevant to have it in the archive though.

@dob71 do you have a use case where `bml_nw_map_query` is useful?

rmelotte committed:
.gitlab-ci.yml: kill prplmesh on all targets before running a test

At the moment every test target are on the same network. If for
whatever reason a prplMesh is running on another target then the one
being tested, it has a high chance of interfering with the test.

We initially wanted to solve this by using VLANs to isolate the
devices. However, because Multi-AP R2 uses VLANs, there will be less
work to do to make the CI setup work with R2 if we avoid using VLANs
at all.

Add a new variable, ALL_TARGETS, which lists all the targets we
currently have in the setup.
Try to kill prplMesh on all the targets in a before_script. If it
doesn't work because prplmesh is not running or because the specified
target is not reachable, do not fail.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -80,10 +96,11 @@ main() {
     dbg "OPENWRT_VERSION=$OPENWRT_VERSION"
     dbg "PRPL_FEED=$PRPL_FEED"
-    dbg "IMAGE_ONlY=IMAGE_ONLY"
+    dbg "IMAGE_ONlY=$IMAGE_ONLY"
```
There was still a typo in the prompt here, so I fixed that as well.

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -48,24 +54,34 @@ build_prplmesh() {
     while true; do
         case "$1" in
             -h | --help)               usage; exit 0; shift ;;
+            -b | --build-options)      BUILD_OPTIONS="$2"; shift ;;
```
There's a missing shift here, so the option will be interpreted as a script option as well...

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -48,24 +54,34 @@ build_prplmesh() {
+			for i in $SUPPORTED_TARGETS ; do
+			    info "$i"
+			done
+                        exit 1
```
Tabs instead of spaces...

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -48,24 +54,34 @@ build_prplmesh() {
                 case "$TARGET_DEVICE" in
                     turris-omnia)
                         TARGET=mvebu
                         ;;
```
This code wasn't executed at all if the -d option is not given, so there is no error in that case. To simplify the code, I just moved this entire case block after the loop. That also makes it more readable IMO.

arnout commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -20,13 +21,18 @@ usage() {
            --build-arg TARGET \
            --build-arg PRPL_FEED \
-           "$scriptdir/"
+           "$ARGS"
```
This isn't going to work: the whole thing will be passed as a single argument with an embedded space and won't be recognized as an option.

I changed this to instead use the BUILD_OPTIONS directly, unquoted (so if multiple options are provided they are properly passed as individual options).

arnout reviewed (approved):
The new -b option didn't work, so I did a bunch of fixes.

There were also a bunch of tabs/spaces indentation errors. Perhaps we should add a [modeline](https://help.gnome.org/users/gedit/stable/gedit-plugins-modelines.html) to all files so editors do the right thing? @tomereli @rmelotte @dob71 @morantr does your editor support modeline? Normal editors (i.e. vim, emacs, kate, gedit) of course do support it, and I know that for Atom you need an [extension](https://atom.io/packages/vim-modeline), but I don't know about weird editors like vscode.
